python/lsst/meas/algorithms/sourceSelector.py

```diff
@@ -97,3 +103,301 @@ def _isBad(self, source):
+                              doc="Select objects with value less than this.")
+
+    def apply(self, catalog):
+        """Apply the flag requirements to a catalog
+
+        Returns whether the source is selected.
+
+        Parameters
+        ----------
+        catalog : `lsst.afw.table.SourceCatalog`
+            Source to which to apply limits.
+
+        Returns
+        -------
+        selected : `numpy.ndarray`
+            Boolean array indicating whether each source is selected.
+        """
+        selected = np.ones(len(catalog), dtype=bool)
+        value = catalog[self.name]
+        if self.minimum is not None:
+            selected &= value > self.minimum
+        if self.maximum is not None:
+            selected &= value < self.maximum
+        return selected
+
+
+class ScienceSourceSelectorConfig(pexConfig.Config):
+    """Configuration for selecting science sources"""
+    doFluxLimit = pexConfig.Field(dtype=bool, default=False, doc="Apply flux limit?")
+    doFlags = pexConfig.Field(dtype=bool, default=False, doc="Apply flag limitation?")
+    doUnresolved = pexConfig.Field(dtype=bool, default=False, doc="Apply unresolved limitation?")
+    fluxLimit = pexConfig.ConfigField(dtype=FluxLimit, doc="Flux limit to apply")
+    flags = pexConfig.ConfigField(dtype=RequireFlags, doc="Flags to require")
+    unresolved = pexConfig.ConfigField(dtype=RequireUnresolved, doc="Star/galaxy separation to apply")
+
+    def setDefaults(self):
+        pexConfig.Config.setDefaults(self)
+        self.flags.bad = ["base_PixelFlags_flag_edge", "base_PixelFlags_flag_interpolated",
+                          "base_PixelFlags_flag_saturated"]
+
+class ScienceSourceSelectorTask(BaseSourceSelectorTask):
+    """Science source selector
+
+    This selects (science) sources by (optionally) applying each of a
+    magnitude limit, flag requirements and star/galaxy separation.
+    """
+    ConfigClass = ScienceSourceSelectorConfig
+
+    def selectSources(self, catalog, matches=None):
+        """Return a catalog of selected sources
+
+        Parameters
+        ----------
+        catalog : `lsst.afw.table.SourceCatalog`
+            Catalog of sources to select.
+        matches : `lsst.afw.table.ReferenceMatchVector`, optional
+            List of matches; ignored.
+
+        Return struct
+        -------------
+        sourceCat : `lsst.afw.table.SourceCatalog`
+            Catalog of selected sources, non-contiguous.
+        """
+        selected = np.ones(len(catalog), dtype=bool)
+        if self.config.doFluxLimit:
+            selected &= self.config.fluxLimit.apply(catalog)
+        if self.config.doFlags:
+            selected &= self.config.flags.apply(catalog)
+        if self.config.doUnresolved:
+            selected &= self.config.unresolved.apply(catalog)
+
+        self.log.info("Selected %d/%d sources", selected.sum(), len(catalog))
+
+        result = type(catalog)(catalog.table)
```

Comment:
What does this line do?

Reply:
`type(catalog)` gives the catalog class, then we're calling it: it's a constructor.

Added a comment to these lines.